app/src/main/java/mozilla/lockbox/view/EditItemFragment.kt

```diff
@@ -49,6 +58,130 @@ class EditItemFragment : BackableFragment(), EditItemDetailView {
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
         setupToolbar(view.toolbar)
+        setTextWatcher(view)
+        setUpKeyboardFocus(view)
+
+        val layouts = arrayOf(view.inputLayoutHostname, view.inputLayoutUsername,
+            view.inputLayoutPassword)
+
+        for(layout in layouts) {
+            layout.hintTextColor = context?.getColorStateList(R.color.hint_edit_text)
+            layout.setHintTextAppearance(R.style.HintText)
+        }
+    }
+
+    private fun setUpKeyboardFocus(view: View) {
+        view.inputHostname.onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
+            if (!hasFocus) {
+                closeKeyboard()
+            }
+        }
+
+        view.inputUsername.onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
+            if (!hasFocus) {
+                closeKeyboard()
+            }
+        }
+
+        view.inputPassword.onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
+            if (!hasFocus) {
+                closeKeyboard()
+                dispatcher.dispatch(ItemDetailAction.TogglePassword(displayed = false))
+            } else {
+                // show the password when it is focused
+                dispatcher.dispatch(ItemDetailAction.TogglePassword(displayed = true))
+            }
+        }
+    }
+
+    private fun setTextWatcher(view: View) {
+        view.inputHostname.addTextChangedListener(
+            buildTextWatcher(
+                view.inputLayoutHostname
+            )
+        )
+        view.inputUsername.addTextChangedListener(
+            buildTextWatcher(
+                view.inputLayoutUsername
+            )
+        )
+        view.inputPassword.addTextChangedListener(
+            buildTextWatcher(
+                view.inputLayoutPassword
+            )
+        )
+    }
+
+    private fun buildTextWatcher(errorLayout: TextInputLayout): TextWatcher {
+        return object : TextWatcher {
+            override fun beforeTextChanged(
+                charSequence: CharSequence,
+                start: Int,
+                count: Int,
+                after: Int
+            ) {
+            }
+
+            override fun onTextChanged(
+                charSequence: CharSequence,
+                start: Int,
+                before: Int,
+                count: Int
+            ) {
+            }
+
+            override fun afterTextChanged(editable: Editable) {
+                val inputText: String? = errorLayout.editText?.text.toString()
+
+                when (errorLayout.id) {
+                    R.id.inputLayoutHostname -> {
+                        // hostname cannot be null
+                        // has to have http:// or https://
```

Comment:
You should also ensure that there are no path segments i.e. https://www.example.com is valid whereas https://www.example.com/ and https://www.example.com/foo are invalid.

Reply:
We decided not to allow editing of this field on desktop for now because users can make changes to it which break autofill so you could also de-prioritize editing of the origin and only allow editing this field for new logins though my comments here about not including path segments or limiting the scheme still applies.